File: server/controllers/interviewController.js
const { ProfileRepository, HistoryRepository } = require('../models/dbStore');
const llmGateway = require('../services/llmGateway');

// @desc    Generate structured interview questions across 5 categories
// @route   POST /api/interview/generate
const generateInterviewPrep = async (req, res, next) => {
  try {
    let { company, role, jobDescription, resumeText } = req.body;

    const profile = await ProfileRepository.findOne({ userId: req.user._id });
    if (profile) {
      if (!company) company = profile.company;
      if (!role) role = profile.role;
      if (!jobDescription) jobDescription = profile.jobDescription;
      if (!resumeText) resumeText = profile.resumeText;
    }

    const targetCompany = company || 'Target Company';
    const targetRole = role || 'Software Engineer';

    const prompt = `
Generate targeted interview questions and suggested answers for a candidate interviewing for the role of "${targetRole}" at "${targetCompany}".
Structure the questions into 5 distinct categories:
1. "technical" (3 questions on core CS, stack, problem solving)
2. "hr" (2 questions on career goals, motivation, salary/fit)
3. "behavioural" (2 questions using STAR format situations)
4. "roleSpecific" (2 questions focused explicitly on ${targetRole} day-to-day duties)
5. "companySpecific" (2 questions tailored to ${targetCompany}'s products, scale, and engineering values)

Return ONLY valid JSON with this exact schema:
{
  "technical": [
    { "question": "...", "difficulty": "Easy|Medium|Hard", "topic": "...", "suggestedAnswer": "..." }
  ],
  "hr": [
    { "question": "...", "difficulty": "Easy|Medium|Hard", "topic": "...", "suggestedAnswer": "..." }
  ],
  "behavioural": [
    { "question": "...", "difficulty": "Easy|Medium|Hard", "topic": "...", "suggestedAnswer": "..." }
  ],
  "roleSpecific": [
    { "question": "...", "difficulty": "Easy|Medium|Hard", "topic": "...", "suggestedAnswer": "..." }
  ],
  "companySpecific": [
    { "question": "...", "difficulty": "Easy|Medium|Hard", "topic": "...", "suggestedAnswer": "..." }
  ]
}
`;

    const context = {
      company: targetCompany,
      role: targetRole,
      jobDescription,
      resumeText,
    };

    const aiResult = await llmGateway.generateAIResponse({
      prompt,
      context,
      jsonMode: true,
      moduleType: 'interview',
      systemPrompt: 'You are a Senior Technical Hiring Manager and Interview Architect specializing in FAANG and top-tier tech interviews.',
    });

    const questionsData = aiResult.parsedJson || llmGateway.generateFallbackResponse('interview', context);

    await HistoryRepository.create({
      userId: req.user._id,
      module: 'Interview Preparation',
      company: targetCompany,
      role: targetRole,
      summary: `Generated interview preparation Q&A for ${targetRole} at ${targetCompany}`,
      metadata: { targetCompany, targetRole },
    });

    res.status(200).json({
      success: true,
      data: questionsData,
      provider: aiResult.provider,
    });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  generateInterviewPrep,
};

File: server/src/main/java/com/nexoffer/controller/InterviewController.java
package com.nexoffer.controller;

import com.nexoffer.model.History;
import com.nexoffer.model.Profile;
import com.nexoffer.model.User;
import com.nexoffer.repository.HistoryRepository;
import com.nexoffer.repository.ProfileRepository;
import com.nexoffer.service.LlmGatewayService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/interview")
public class InterviewController {

    private final ProfileRepository profileRepository;
    private final HistoryRepository historyRepository;
    private final LlmGatewayService llmGatewayService;

    public InterviewController(ProfileRepository profileRepository,
                               HistoryRepository historyRepository,
                               LlmGatewayService llmGatewayService) {
        this.profileRepository = profileRepository;
        this.historyRepository = historyRepository;
        this.llmGatewayService = llmGatewayService;
    }

    @PostMapping("/generate")
    public ResponseEntity<?> generateInterviewPrep(@AuthenticationPrincipal User user, @RequestBody(required = false) Map<String, String> body) {
        String company = body != null ? body.get("company") : null;
        String role = body != null ? body.get("role") : null;
        String jobDescription = body != null ? body.get("jobDescription") : null;
        String resumeText = body != null ? body.get("resumeText") : null;

        Optional<Profile> profileOpt = profileRepository.findByUserId(user.getId());
        if (profileOpt.isPresent()) {
            Profile p = profileOpt.get();
            if (company == null || company.isBlank()) company = p.getCompany();
            if (role == null || role.isBlank()) role = p.getRole();
            if (jobDescription == null || jobDescription.isBlank()) jobDescription = p.getJobDescription();
            if (resumeText == null || resumeText.isBlank()) resumeText = p.getResumeText();
        }

        String targetCompany = (company != null && !company.isBlank()) ? company : "Target Company";
        String targetRole = (role != null && !role.isBlank()) ? role : "Software Engineer";

        String prompt = "Generate targeted interview questions and suggested answers for a candidate interviewing for the role of \"" + targetRole + "\" at \"" + targetCompany + "\".\n" +
                "Structure the questions into 5 distinct categories: technical (3 Qs), hr (2 Qs), behavioural (2 Qs), roleSpecific (2 Qs), companySpecific (2 Qs).\n" +
                "Return ONLY valid JSON with this schema:\n" +
                "{\n" +
                "  \"technical\": [{ \"question\": \"...\", \"difficulty\": \"Easy|Medium|Hard\", \"topic\": \"...\", \"suggestedAnswer\": \"...\" }],\n" +
                "  \"hr\": [{ \"question\": \"...\", \"difficulty\": \"Easy|Medium|Hard\", \"topic\": \"...\", \"suggestedAnswer\": \"...\" }],\n" +
                "  \"behavioural\": [{ \"question\": \"...\", \"difficulty\": \"Easy|Medium|Hard\", \"topic\": \"...\", \"suggestedAnswer\": \"...\" }],\n" +
                "  \"roleSpecific\": [{ \"question\": \"...\", \"difficulty\": \"Easy|Medium|Hard\", \"topic\": \"...\", \"suggestedAnswer\": \"...\" }],\n" +
                "  \"companySpecific\": [{ \"question\": \"...\", \"difficulty\": \"Easy|Medium|Hard\", \"topic\": \"...\", \"suggestedAnswer\": \"...\" }]\n" +
                "}";

        Map<String, String> context = new HashMap<>();
        context.put("company", targetCompany);
        context.put("role", targetRole);
        if (jobDescription != null) context.put("jobDescription", jobDescription);
        if (resumeText != null) context.put("resumeText", resumeText);

        LlmGatewayService.AiResponse aiResponse = llmGatewayService.generateAIResponse(
                prompt,
                "You are a Senior Technical Hiring Manager and Interview Architect specializing in FAANG and top-tier tech interviews.",
                context,
                true,
                "interview"
        );

        Object data = aiResponse.getParsedJson() != null ? aiResponse.getParsedJson() : llmGatewayService.generateFallbackResponse("interview", context);

        historyRepository.save(new History(
                user.getId(),
                "Interview Preparation",
                targetCompany,
                targetRole,
                "Generated 5-category interview questions"
        ));

        return ResponseEntity.ok(Map.of(
                "success", true,
                "data", data,
                "provider", aiResponse.getProvider()
        ));
    }
}
